lib/api.js

```diff
@@ -11,35 +11,80 @@ const ApiError    = require('./errors/api-error');
 
 const httpsProxyAgent = require('https-proxy-agent');
 
-function _responseHandler(req) {
+const DEFAULT_MAX_NETWORK_RETRY = 0;
+const MAX_NETWORK_RETRY_DELAY = 2000; // in ms
+const INITIAL_NETWORK_RETRY_DELAY = 500; // in ms
+
+function _processHttpRequest(
+  requestOptions,
+  maxNetworkRetries,
+  totalRetry = 0
+) {
   return new Promise(function(resolve, reject) {
+    const scheme = process.env.OMISE_SCHEME || requestOptions.scheme;
+    const req = scheme == 'http'
+      ? http.request(requestOptions)
+      : https.request(requestOptions);
+
+    if (_requestMethodWithBody(requestOptions.method)) {
+      req.write(requestOptions['body'], 'utf8');
+    }
+
     req.on('response', function(res) {
-      let resp = '';
-      res.setEncoding('utf8');
-      res.on('data', function(chunk) {
-        resp += chunk;
-      });
-      res.on('end', function() {
-        try {
-          logger.log('info', resp);
-          if (!_isValidJsonString(resp)) {
-            reject(resp);
-          }
-          resp = JSON.parse(resp);
-          if (resp.object === 'error') {
-            reject(resp);
-          } else {
-            resolve(resp);
-          }
-        } catch (err) {
-          reject(err);
-        }
-      });
-    })
-      .on('error', reject);
+      _handleHttpResponse(res, resolve, reject);
+    }).on('error', (err) => {
```

Comment:
Which error cases does this cover?

Reply:
- Network-related errors
- Connection errors
- HTTP protocol errors and 
- Server errors(status code in the range of 500-599)

Reply:
Can you point me to the docs?

Reply:
https://nodejs.org/api/http.html#http_http_request_options_callback

Reply:
Our APIs handles 500 and 503 errors and responds a JSON. This will not cover those cases, will it? Correct me if I am wrong.

https://www.omise.co/api-errors#500-internal-server-error